[WIP] Hash table unification

lorentey opened the pull request:
- Extract low-level hash table operations into a type-agnostic `_HashTable` struct, parts of which can potentially be made resilient.
- Store 7 additional bits of the hash value directly in each entry's metadata field. (The hope is that this will allow us to use a much higher max load factor, by making collision chains dramatically cheaper.)
- Updated storage class hierarchy and internal storage layout for `Set` and (soon) `Dictionary`.
- New representation for the type-punned empty singleton storage.
- Slightly less memory overhead for deferred bridging. (This is especially relevant for `Dictionary` where only one of the element types may need non-verbatim bridging.)

This hasn't been fully debugged yet, and also needs to be extensively benchmarked/fine-tuned.

An lldb data formatter update will need to land before merging this PR.

lorentey committed:
[stdlib] Add struct _HashTable

This implements low-level hash table operations behind a (partially) resilient interface, allowing us to potentially not expose hash table metadata layout as part of the stdlib ABI.

lorentey committed:
[stdlib] Improve Set storage layout

- Separate low-level hash table operations into a new _HashTable struct. (Only used by Set for now.) This enables us to hide some aspects of hash tables behind a resilience boundary.
- Cache seven bits from the hashValue in hash table metadata. This makes collisions far cheaper, paving the way to increasing the load factor.
- Introduce a separate class for the empty set singleton (_SwiftEmptySetStorage). Using the raw storage superclass for this was problematic, as it sometimes confused lldb.
- Get rid of the non-Hashable native storage class. Introduce a trivial _BridgingHashBuffer class for use within _SwiftDeferredNSSet.

- Rename storage classes. The new names make it easier for lldb to use the correct formatter logic, and they make for easier recognition in heap dumps etc.
    _RawNativeSetStorage => _SwiftRawSetStorage
    _HashableTypedNativeSetStorage => _SwiftNativeSetStorage
    _SwiftDeferredNSSet => _SwiftDeferredNSSet

lorentey committed:
[stdlib] Make _stdlib_NSObject_isEqual internal

lorentey committed:
[stdlib] _makeCollectionDescription(for:,withTypeName:): make internal

lorentey committed:
[REMOVE][stdlib] Set._dump(): Temporarily add to aid in debugging

lorentey committed:
[stdlib] _HashTable: Make more members internal

phausler commented on stdlib/public/core/Set.swift:
```diff
@@ -1503,163 +1505,208 @@ internal protocol _SetBuffer { // FIXME: Remove or refactor for Set.
-    _sanityCheckFailure("Only create this by using the `empty` singleton")
-  }
 
+extension _SwiftEmptySetStorage: _NSSetCore {
```
FYI we did find it worthwhile in Foundation to have 1 element variants as well in addition to this strategy of a zero element container.

Couldn't this be a singleton?

phausler commented on stdlib/public/core/Set.swift:
```diff
@@ -2013,259 +2016,268 @@ extension _NativeSet/*: _SetBuffer*/ where Element: Hashable {
+  fatalError("""
+    Duplicate elements of type '\(elementType)' were found in a Set.
+    This usually means that either the type violates Hashable's requirements, or
+    that members of such a set were mutated after insertion.
```
the case we have seen often in Foundation is where an element of a NSMutableSet has a non-stable hash value. is this function breakpointable to help developers figure out why they are getting this failure?

lorentey replied:
While the function is declared internal, the `@usableFromInline` attribute means there will still be a symbol generated for it -- so it'll be breakpointable. (It essentially makes the symbol public on the ABI level, so lldb will see it.)

I'll have to play with this a little to make sure the idea is sound. Multiline error messages may cause issues. Also, I think `fatalError` generates a bunch of stack frames before aborting, so I'm not sure if this will be as visible as I hope. Plus I think it may be useful to print one of the duplicate values (passed in as an `Any` parameter), along with the type -- although that may be a step too far.

lorentey committed:
[stdlib] Set.update(with:): optimize storage handling

lorentey committed:
[test] Update names of native Set storage classes

lorentey committed:
[runtime] Update layout of empty set singleton instance

lorentey committed:
[stdlib] _HashTable: Introduce capacity(forScale:) and _idealBucket(forHashValue:)

lorentey committed:
[REMOVE][stdlib] Add Set._invariantCheck()

lorentey committed:
[REMOVE][stdlib] Add Set._stats

lorentey committed:
[test] Update empty case in Set.removeAll tests

We now discard storage even if the element count is zero. (This is the correct behavior; the test was wrong.)

lorentey commented:
@swift-ci benchmark

lorentey committed:
[stdlib] Set(_cocoa:): New initializer

Don’t call init(_immutableCocoaSet:) to construct a Set from a _CocoaSet.

lorentey committed:
[stdlib] Set._Variant: Replace static factories with initializers

lorentey committed:
[stdlib] Update comments

lorentey committed:
[stdlib] _Bitmap: New value type representing a fixed-size bitmap

This is intended to replace _UnsafeBitmap; however, that is currently still being used by Dictionary.

lorentey committed:
[stdlib] Generalize _NativeSet.insertNew(_:) and use in more places

Also, rename moveNew to _unsafeMoveNew.
Build sets using _NativeSet.insertNew(_:) in these places:

  Set.init<S: Sequence>(_:)
  Set.filter(_:)
  _setBridgeToObjectiveC<S, O>(_:)
  _setDownCastConditional<B, D>(_:)
  _setBridgeFromObjectiveCConditional<O, S>(_:)

This is now safe because insertNew maintains a valid set and handles resizing as necessary.

lorentey committed:
[stdlib] Set: Optimize isSubset/isSuperset tests

- Test for trivial fast paths.
- Don’t hash the input sequence more than once — mark elements seen in a temporary bitmap and work from that.

lorentey committed:
[stdlib] Set: Unify isDisjoint(_:) variants.

lorentey committed:
[stdlib] Set: Optimize intersection

- Use bitmaps to eliminate the need for multiple passes on the elements of the input sequence.
- When given two sets, prefer calculating the intersection on the native one. When both sets are native, iterate over the one with fewest elements.

lorentey committed:
[stdlib] Set: Optimize symmetric difference

- Use bitmaps to eliminate the need for multiple passes on the elements of the input sequence.
- When given two sets, prefer calculating the intersection on the native one. When both sets are native, iterate over the one with fewest elements.

lorentey committed:
[stdlib] Set: Optimize formUnion

- The union operation is odd because it seems best to base it on its mutating variant.
- When given the option, start with a native set. This may prevent one round of rehashing.
- When both sides are sets, they establish a minimum bound on the result's element count.

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please benchmark

lorentey commented:
TEST                                                      | OLD    | NEW    | DELTA  | SPEEDUP   
---                                                       | ---    | ---    | ---    | ---       
SetIntersect                                              | 587    | 1022   | +74.1% | **0.57x** 
SetUnion                                                  | 4148   | 6054   | +45.9% | **0.69x** 

That's funny! These results are the opposite of my own `Set<Int>` benchmarking.

lorentey commented:
![set - 4 tasks](https://user-images.githubusercontent.com/608696/44469358-3a78ea80-a61f-11e8-9be7-ccae77ffff6e.png)

Separately benchmarking SetIntersect in detail reveals an important clue -- the benchmark as it currently stands measures the intersection between two non-overlapping sets. I can't locally reproduce the +75% slowdown -- this use-case benchmarks roughly the same before/after this PR.

Fixing the benchmark to include a non-empty intersection (with ~25% of elements in both sets) shows clear improvements over the non-overlapping case. (See `SetIntersection2` above.) This is rather surprising, as the algorithm actually needs to do *more* work in the overlapping case, not less.

lorentey commented:
`SetIntersection` uses a random number generator, which generates duplicate elements in the `SetIntersection2` case, making `SetIntersection` and `SetIntersection2` not directly comparable. I submitted #18928 to fix these benchmarks on master.

Updated chart:
![set - 4 tasks](https://user-images.githubusercontent.com/608696/44549421-ee11d580-a718-11e8-8cee-8003cc5f247a.png)

Note how the new algorithm is faster at almost every stage. It gets relatively faster as the % of common elements increase. The intersection of two equal sets is calculated particularly efficiently:

![set - 4 tasks](https://user-images.githubusercontent.com/608696/44549596-7b552a00-a719-11e8-9238-beb66bd4c7cc.png)

The new algorithm detects that there is no need to construct a third set, making it even faster than the 0% case for many sizes. (This is because looking up an existing element is faster than one that's not in the hash table.)

lorentey committed:
[stdlib] _HashTable.scale: expose to inlinable code

Exposing the number of buckets allows _NativeSet.reallocate() to be made inlinable, which seems to be important for performance.

lorentey commented:
@swift-ci please test linux platform

lorentey commented:
@swift-ci please test linux platform

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey committed:
[stdlib] _Bitmap: Apply review notes

lorentey committed:
[stdlib] Set: optimize subtraction

lorentey commented:
Subtraction optimizations in the latest updates should help clean up benchmark regressions.

@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey committed:
[squash] Fix isStrictSuperset oversight

lorentey committed:
[stdlib] Hasher._Seed: New typealias

I don’t expect we’ll ever need to change this, but the typealias is a readability win.

lorentey committed:
[stdlib] Set: Enable true per-instance seeding

Set.union and .symmetricDifference still exhibit quadratic behaviour with similar-sized but almost disjunct input sets.

This was due to hash values correlation between hash tables of the same size. Enabling true per-instance seeding fixes this.

lorentey committed:
[stdlib] _HashTable.MapEntry: Add explicit conformance to Equatable

lorentey committed:
[stdlib] Set: Remove guaranteedNative fastPath

Evidently it has no effect on benchmarks.

lorentey committed:
[squash][stdlib] _SwiftRawSetStorage.seed: Make final

Leaving this overridable caused a ~2x slowdown on Set<Int>.contains benchmarks. D’oh.

lorentey committed:
[squash][stdlib] HashTable: Reorder stored properties; remove count from initializer

lorentey committed:
[squash][stdlib] _HashTable.payloadMask, .unoccupied: Force inlining

lorentey committed:
[squash][stdlib] _HashTable: More force-inlining

lorentey committed:
[stdlib] _NativeSet.find: Small speedup

lorentey committed:
[stdlib] _HashTable: make bucketCount a stored property

lorentey committed:
[stdlib] _HashTable: Make everything inlinable

This gets us a nice baseline to start with.

lorentey commented:
Latest commits get us to master's performance or better for most benchmarks, at the cost of disabling hash table layout resilience. 

These commits have not yet shown up in this PR; let's see if CI picks them up anyway.

@swift-ci please smoke benchmark staging

lorentey commented:
I see GitHub has caught up now.

@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please benchmark

lorentey commented:
@swift-ci please benchmark

lorentey committed:
[remove][stdlib] More statistics

lorentey committed:
[squash][stdlib] Report Hashable violations with _assertionFailure

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey committed:
[stdlib] _HashTable: Rewrite applying some SIMD(ish) tricks

- Reduce the number of hash table buckets by a factor of 8.
- Increase the size of buckets from 1 to 8 entries.
- Use SIMD within a register to find matching entries, instead of iterating over individual bytes.

lorentey committed:
[stdlib] Speed up Set.contains

- Simplify generated code; eliminate obvious inefficiencies.
- Revive guaranteedNative trick; removing the bridged branch does have a measurable impact for contains at least. (Apart from code size benefits)

Lookups in sets are still not quite as fast as on master for small sets (up to ~1k elements) when the element type has fast equality checks — I believe this is because the SIMD approach has worse latency than direct lookup, and the lookup chains aren’t long enough to take advantage of higher throughput. (The latency difference only matters when the set is fully cached, which is why the slowdown only occurs for small sets.)

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging
